torch/csrc/autograd/FunctionsManual.cpp

```diff
@@ -2491,6 +2491,42 @@ Tensor linalg_eig_backward(const std::vector<torch::autograd::Variable> &grads,
   }
 }
 
+// https://people.maths.ox.ac.uk/gilesm/files/NA-08-01.pdf, page 10
+// see also https://arxiv.org/pdf/1701.00392.pdf Eqs. (4.60) and (4.63)
+Tensor linalg_eig_jvp_eigenvalues(const Tensor& dA,
+                                  const Tensor& L,
+                                  const Tensor& V) {
+  const auto dAComplex = at::complex(dA, at::zeros(dA.size(-1), dA.options()));
+  const auto dVfactor = at::linalg_solve(V, at::matmul(dAComplex, V));
+  return at::diag(at::mul(at::eye(V.size(-1), V.options()), dVfactor));
+}
+
+Tensor linalg_eig_jvp_eigenvectors(const Tensor& dA,
+                                   const Tensor& L,
+                                   const Tensor& V) {
+  const auto dAComplex = at::complex(dA, at::zeros(dA.size(-1), dA.options()));
+  const auto dVfactor = at::linalg_solve(V, at::matmul(dAComplex, V));
+  const auto Lconj = L.conj();
+  const auto Econj = Lconj.unsqueeze(-2) - Lconj.unsqueeze(-1);
+  auto Fconj = at::pow(Econj, -1);
+  Fconj.diagonal(0, -2, -1).zero_();
+  return at::matmul(V, at::mul(Fconj, dVfactor));
+}
+
+std::tuple<Tensor, Tensor> linalg_eig_jvp(const Tensor& dA,
+                                          const Tensor& L,
+                                          const Tensor& V) {
+  const auto dAComplex = at::complex(dA, at::zeros(dA.size(-1), dA.options()));
+  const auto dVfactor = at::linalg_solve(V, at::matmul(dAComplex, V));
+  const auto Lconj = L.conj();
+  const auto Econj = Lconj.unsqueeze(-2) - Lconj.unsqueeze(-1);
+  const auto Fconj = at::pow(Econj, -1);
```

Comment:
While the paper does it this way for notational convenience, it's better to just divide later when you use it.

Reply:
> Sorry for coming late to the party!
> 
> I left a few comments.
> 
> As the code is repeated three times, the changes should be applied in the 3 places. Even better, it would be to have the two auxiliary functions just call the main function that returns the pair and `std::get` the appropriate return.

I was thinking about factoring the code like that too, but then I thought "this is only provisional until the codegen limitation is fixed". But now it seems like we use the separate calls for eigenvalues and eigenvectors. Anyway, I'll fix it like you suggested it.

Reply:
For what is worth, we're working on always enforcing to compute the `jvp` for all the multiple outputs at once and returning them in a tuple, so having just one function and then using the `std::get`s is the best future-proof solution :)